### ISSUE TITLE
haproxy -sf <pid> is unable to recover from errors in previous haprx -sf invocations

### ISSUE THREAD

**Issue by geofffranks** [haproxy/haproxy#1883]
### Detailed Description of the Problem

We reload haproxy frequently when configs change via `haproxy -f <config> -p <pidfile> -x <socket_file> -D -sf $(cat <pidfile>)`. If a user provides bad configuration to us, and we try to reload haproxy with a config that does not work (e.g. the port is already bound), haproxy is unable to reload *after* the config problem has resolved.

### Expected Behavior

After fixing the configuration to be something that does not use in-use ports, or killing what was in-use, `haproxy -sf $(cat <pidfile>) -x <socket-file>` should be able to reload successfully.

### Steps to Reproduce the Behavior

1. Launch HAProxy with a config that listens on port 18000.
2. Update the haproxy config to add port 18001 as another listener
3. Run `nc -l 18001`
4. Reload haproxy via `haproxy -f <config> -p <pidfile> -x <socketfile> -D -sf $(cat <pidfile>)`
5. It should fail indicating the port is in use (expected and correct behavior):
```
[NOTICE]   (14531) : haproxy version is 2.6.6-274d1a4
[NOTICE]   (14531) : path to executable is /var/vcap/packages/haproxy/bin/haproxy
[ALERT]    (14531) : Binding [/var/vcap/data/tcp_router/config/haproxy.conf:78] for frontend frontend_18001: cannot bind socket (Address already in use) for [0.0.0.0:18001]
[ALERT]    (14531) : [/var/vcap/packages/haproxy/bin/haproxy.main()] Some protocols failed to start their listeners! Exiting.
```
6. Kill the `nc` process listening on 18001.
7. Reload haproxy via `haproxy -f <config> -p <pidfile> -x <socketfile> -D -sf $(cat <pidfile>)`
8. It will fail with the following:
```
[NOTICE]   (14535) : haproxy version is 2.6.6-274d1a4
[NOTICE]   (14535) : path to executable is /var/vcap/packages/haproxy/bin/haproxy
[WARNING]  (14535) : Failed to connect to the old process socket 'haproxy.sock'
[ALERT]    (14535) : Failed to get the sockets from the old process!
```

### What is your configuration?

```haproxy
global
    log 127.0.0.1   syslog info
    daemon
    maxconn 64000
    spread-checks 4
    stats socket /var/vcap/data/tcp_router/config/haproxy.sock user vcap group vcap mode 600 level operator expose-fd listeners

defaults
    log global
    timeout connect 30000ms
    maxconn 64000

        timeout client 300000ms
        timeout server 300000ms


listen health_check_http_url
    mode http
    bind :80
    monitor-uri /health

frontend frontend_18000
  mode tcp
  bind :18000
  default_backend backend_18000

backend backend_18000
  mode tcp
  server server_10.0.4.8_61000 10.0.4.8:61000

frontend frontend_18001
  mode tcp
  bind :18001
  default_backend backend_18001

backend backend_18001
  mode tcp
  server server_10.0.4.8_61000 10.0.4.8:61000
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.6-274d1a4 2022/09/22 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.6.html
Running on: Linux 4.15.0-192-generic #203~16.04.1-Ubuntu SMP Fri Aug 12 09:07:51 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment -fcommon
  OPTIONS = USE_OPENSSL=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL -LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=1).
Built with OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
Running on OpenSSL version : OpenSSL 1.0.2g  1 Mar 2016
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 5.4.0 20160609

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Additional Information

In 1.8.13, it errors when the port is in use, but then succeeds afterwards. We upgraded to 2.5.4 and noticed the problem, and then upgraded to 2.6.6, and the problem persists.

**Comment by wlallemand** [comment]
I couldn't reproduce the behavior you are facing with any of the 1.8.13, 1.8.30 or 2.6.6 version.

```
[WARNING]  (14535) : Failed to connect to the old process socket 'haproxy.sock'
[ALERT]    (14535) : Failed to get the sockets from the old process!
```

This error means the new process was unable to connect to the socket of the previous one. I checked the commits and it always generated an alert since we added this feature.

Are you sure you are using the right parameter for -x ? Because it says "haproxy.sock" instead of /var/vcap/data/tcp_router/config/haproxy.sock

**Comment by geofffranks** [comment]
Yeah, I was in the `/var/vcap/data/tcp_router/config` directory when I ran it, and forgot to append the $PWD variable, but the results are the same either way.

We spent some time today on more reproduction steps outside of our environment, in the haproxy docker images. We did run into issues reproducing in the docker images, where it was able to bind to the port-already-in-use  if the same user had created the netcat process. However, when that tcp socket was owned by a different user, we were able to reproduce.

Can you try the steps/docker image/configs located on [this gist](https://gist.github.com/geofffranks/3518cdc3b6087fd67ef885825dc52a7b)?

Some interesting things we noticed while testing:
1. when reloading haproxy and it fails due to port-in-use, the inode of the socket file changes.
2. strace showed the following output related to the socket file, which made it look like maybe it was replacing the socket file before everything had successfully started up? When running strace against 1.8.13, we did not see anything other than the initial connect() and getsockname() calls in strace.

```
connect(4, {sa_family=AF_LOCAL, sun_path="/var/vcap/data/tcp_router/config/haproxy.sock"}, 110) = 0
getsockname(12, {sa_family=AF_LOCAL, sun_path="/var/vcap/data/tcp_router/config/haproxy.sock.20366.tmp"}, [58]) = 0
unlink("/var/vcap/data/tcp_router/config/haproxy.sock.20383.tmp") = -1 ENOENT (No such file or directory)
unlink("/var/vcap/data/tcp_router/config/haproxy.sock.20383.bak") = -1 ENOENT (No such file or directory)
link("/var/vcap/data/tcp_router/config/haproxy.sock", "/var/vcap/data/tcp_router/config/haproxy.sock.20383.bak") = 0
bind(4, {sa_family=AF_LOCAL, sun_path="/var/vcap/data/tcp_router/config/haproxy.sock.20383.tmp"}, 110) = 0
chown("/var/vcap/data/tcp_router/config/haproxy.sock.20383.tmp", 1000, 1000) = 0
chmod("/var/vcap/data/tcp_router/config/haproxy.sock.20383.tmp", 0600) = 0
rename("/var/vcap/data/tcp_router/config/haproxy.sock.20383.tmp", "/var/vcap/data/tcp_router/config/haproxy.sock") = 0
unlink("/var/vcap/data/tcp_router/config/haproxy.sock.20383.bak") = 0
```

**Comment by Darlelet** [comment]
Thank you for providing the docker reproducer.
It also affects 2.7.

> strace showed the following output related to the socket file, which made it look like maybe it was replacing the socket file before everything had successfully started up?

You assumption is right, after config is successfully parsed, haproxy tries to run protocol_bind_all(), but the issue here is that even if one or multiple binds for the frontend sections are failing, stats socket bind is still happening.

In addition it is worth mentioning that underlying unix socket binding function (sock_unix_bind_receiver) is replacing the socket file with its own (unlink on your strace output) when invoked.

To sum up, we are currently witnessing the following (simplified) sequence:
```
config parse: OK
err = ERR_NONE
protocol_bind_all:
   bind frontend_18000: OK
   bind frontend_18001: NOK
         err|=ERR_RETRYABLE
   bind global.stats_socket: OK (replaces old process sock file)
if (err != ERR_NONE)
    "Some protocols failed to start their listeners! Exiting."
    exit
/* ... */
kill old process (SIGUSR1 for "sf" or SIGTERM for "st")
```

protocol_bind_all() keeps iterating through protocol receivers/listeners as long as
ERR_ABORT is not met, but here failing binds for frontend listeners only returns ERR_RETRYABLE: the loop is not interrupted.

It is that way because haproxy will try to run protocol_bind_all() multiple times (up to `MAX_START_RETRIES`, currently set to 200) until no more errors are returned or fatal (non recoverable) error is met, to greatly increase the chances of success.
ERR_RETRYABLE means that the bind currently fails, but the next try could be successful (could be temporary resource shortage/unavailability)

This issue does not only affect "-sf" + "-x" mode, because using "-sf" or "-st" with a new process failing at protocol_bind_all() would be enough to break old process stats socket (existing connections are not affected).

I don't know how to solve this issue, but maybe we could pass an extra variable to protocol_bind_all() that indicates that we're in "sf or st" mode, and in such case, we could force the function to fail as soon as an error is met to prevent stats socket bind from happening and let the retry responsibility to the process manager?

Yet I'm not satisfied with this solution because to my knowledge, protocols list is not explicitly ordered, is it?
(family registration is performed through INITCALLS)
Because of that, I guess it's not safe to assume that global (stats socket) receiver will always be at the end of protocol's list when protocol_bind_all() gets executed?

**Comment by Darlelet** [comment]
Thank you as well for mentioning that it used to work with 1.8.

This regression has been introduced with: e91bff213485e495a12f7576e39f423b3fe10244

Up to 2.3 it "used to work" (I'm quoting as I don't know if it was intended) because proxy listeners were initialized first separately from the protocols:

frontend bind error would show up during proxy listener initialization and abort process startup
(before protocol_bind_all() is actually executed, which handles unix stat socket initialization)

Starting with e91bff213485e495a12f7576e39f423b3fe10244, both proxy listeners and protocols are handled simultaneously by protocol_bind_all(), there is no "pre" proxy listeners initialization in startup path.

**Comment by wtarreau** [comment]
Thanks for the analysis. What made the problem visible is indeed that there is accidental ordering there.

The problem with temporary unbinding of a UNIX socket is that t cannot be done without FS access, and that only the new process could deal with this, which is why it already starts in two steps, first binds a temporary one, then renames it. We've changed a bit the binding sequence over time since 1.5 or so, but it roughly remained the same.

I'm seeing two possible approaches here:
  - either the new process renames the existing socket (eg. $sock.$pid) to preserve it until it starts, and renames it back on failure
  - or we use a two-phase commit for binding the protocols. This would mean that the first call would just bind all sockets, and the second one would commit these sockets (i.e. rename the sockets).

The first one might be more robust because there's no risk of failing to finish (it's just a cleanup phase once everything is properly bound). But it can be a bit more difficult to implement and could possibly trigger issues on long path names (unix sockets are limited to 128 chars so we must be careful about not adding too many chars). We could implement numbered renames instead (i.e. append ".1", ".2", etc).

The two-phase commit could be more resilient to parallel reloads but these are not supposed to happen, and if one fails at the last moment there is no rollback anymore.

A combination of both would also be possible, i.e. 3-phase start:
  - bind all sockets, unix uses .tmp
  - rename all existing unix sockets to .$number, and .tmp to final name
  - cleanup: remove all .$number

In case of any error during the 2nd phase, the rollback is feasible. It's just much more complicated with little perceived benefit.

It would be nice to try to address this before 2.7 so that we can see if we end up with something backportable or not.

**Comment by wtarreau** [comment]
By the way there might be something trickier that happens when -x is used. I wouldn't be surprised that we directly use the socket's FD and that as a result we don't keep a rename of the existing one, but still remove it upon exit (to be verified). If so, we should at least refrain from removing on-disk sockets for those that were retrieved via FD passing!

**Comment by wtarreau** [comment]
Just checked the code and we already take care of not removing inherited sockets on error, but they're definitely done on success. However there's already everything there for a two-step approach if needed, the cleanup part just needs to be placed into a separate function, and we could have the rollback code separate as well for the case it fails (see sock_unix_bind_receiver()).

**Comment by Darlelet** [comment]
Thank you Willy for those interesting leads, just to let you know that I'm working on it!

**Comment by Darlelet** [comment]
Here is a patch that should address this issue, implementing 2phase commit as suggested by Willy.
[0001-MINOR-protocol-implement-2-steps-protocol-bind-w-com.patch.txt](https://github.com/haproxy/haproxy/files/10057396/0001-MINOR-protocol-implement-2-steps-protocol-bind-w-com.patch.txt)

I added some notes if 3phase commit (using additional cleanup+revert handlers) would come to be relevant.

Edit: please don't mind this one, I will quickly re-upload it with a minor fix.

**Comment by Darlelet** [comment]
Here is the final patch, I guess it's a bit cleaner than the previous one because it does not require RX_F_COMMIT flag anymore.
[0001-MINOR-protocol-implement-2-steps-protocol-bind-w-com.patch.txt](https://github.com/haproxy/haproxy/files/10057608/0001-MINOR-protocol-implement-2-steps-protocol-bind-w-com.patch.txt)

**Comment by Darlelet** [comment]
Here is the final patch, I guess it's a bit cleaner than the previous one because it does not require RX_F_COMMIT flag anymore.
[0005-MINOR-protocol-implement-2-steps-protocol-bind-w-com.patch.txt](https://github.com/haproxy/haproxy/files/10057795/0005-MINOR-protocol-implement-2-steps-protocol-bind-w-com.patch.txt)

**Comment by capflam** [comment]
Just to let you know. The patch is too sensitive to be merged at the last minute in the 2.7. The 2.7.0 will be released very soon and we try hard to stabilize it. The patch will be reviewed after that.

**Comment by geofffranks** [comment]
Hey @capflam, any updates on when this might get merged/released?

**Comment by capflam** [comment]
Several patches related to your issue were merged in 2.8-DEV. So it should be fixed. the patches will be backported soon I guess.

**Comment by Darlelet** [comment]
Hi @geofffranks, despite the proposed fix that "seemed" complete this is still a work in progress. 

By digging around we found multiple issues in the same area.

The proposed fix fails to properly address all of them, which explains why it was not merged.

However, as mentioned by @capflam, some work was already carried out:

From your reproducer examples, I see that you're using the `-x` option to retrieve the sockets of the previous process:
There was a bug affecting the retrieval of named UNIX sockets with the `-x` option, and because of this bug
the new process was "forced" to re-create the UNIX sockets (which implies overwriting the previous UNIX sockets files).
This bug was fixed with https://github.com/haproxy/haproxy/commit/2a7903bbb2102867132d9821913c51cb1b938962

While it does not address the root cause (new process that fails to restore the previous UNIX socket paths on errors), it should now work for you since the new process will inherit from the old process sockets without the need to recreate them.
If the new process fails to fully start, the original process will get back to work as if "nothing happened".

However if you remove the `-x` and only rely on the `-sf` you will still encounter the original issue:
if you don't mind we'll keep the issue open until the remaining issues are addressed as well :)
(hopefully this shouldn't take too long)
